### PR TITLE
Rel/3.3.0

### PR DESCRIPTION
**Pull request by alhardy**
- Adding API template with NSB integration
- using health check middleware rather than custom controller
- using health check ui package for health check response formatting

### CODE REVIEW

File: src/ProspaAspNetCoreApiNsb/docs.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿{
+  "Name": "ProspaAspNetCoreApi",
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/docs.json]: I've never seen anyone actually bother to fill this docs.json file in properly

File: src/ProspaAspNetCoreApiNsb/Startup.Versioning.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    public static class StartupVersioning
+    {
+        public static IServiceCollection AddDefaultApiVersioning(this IServiceCollection services)
+        {
+            services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<MvcOptions>, ConventionsMvcOptionsSetup>());
+
+            services.AddApiVersioning(
+                options =>
+                {
+                    options.ApiVersionReader = new UrlSegmentApiVersionReader();
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Startup.Versioning.cs]: why don't we just stick with the default apiVersion Reader?

File: src/ProspaAspNetCoreApiNsb/Startup.SecurityHeaders.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// ReSharper restore CheckNamespace
+{
+    public static class StartupSecurityHeaders
+    {
+        public static IApplicationBuilder UseDefaultSecurityHeaders(this IApplicationBuilder app, IHostEnvironment hostingEnvironment)
+        {
+            if (!hostingEnvironment.IsDevelopment())
+            {
+                app.UseHsts(options => options.MaxAge(days: Constants.HttpHeaders.HstsMaxAgeDays).IncludeSubdomains().Preload().AllResponses());
+                app.UseXfo(options => options.SameOrigin());
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Startup.SecurityHeaders.cs]: I don't think  XFO options makes sense for API projects.

File: src/ProspaAspNetCoreApiNsb/Startup.Mvc.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            services.AddSingleton<IActionContextAccessor, ActionContextAccessor>()
+                    .AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+
+            return services;
+        }
+
+        public static IMvcCoreBuilder AddDefaultMvcOptions(this IMvcCoreBuilder builder)
+        {
+            builder.Services.AddSingleton<IConfigureOptions<MvcOptions>, FormattersMvcOptionsSetup>();
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Startup.Mvc.cs]: can we just define the options inline with a lambda? This extra class just creates an unnecessary level of indirection.

File: src/ProspaAspNetCoreApiNsb/Program.Configuration.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+            if (string.IsNullOrEmpty(keyVaultName))
+            {
+                throw new ApplicationException("A Keyvault name but be present in application settings and the matching Keyvault resource needs to exist. Use az login to authenticate with Keyvault with MSI, ensure that you have permissions to list Keyvault keys");
+            }
+
+            var keyVaultEndpoint = $"https://{Constants.Environments.Prefix()}{keyVaultName}.vault.azure.net/";
+            var azureServiceTokenProvider = new AzureServiceTokenProvider();
+            var keyVaultClient = new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(azureServiceTokenProvider.KeyVaultTokenCallback));
+            builder.AddAzureKeyVault(keyVaultEndpoint, keyVaultClient, new DefaultKeyVaultSecretManager());
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Program.Configuration.cs]: I don't think we need this entire file.
We just want to add keyvault; json, command line and environment variables are already added by default.
IF we define `keyVaultName` as a constant in the code then it is much simpler.

File: src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        private static readonly Lazy<JsonSerializerSettings> Lazy =
+            new Lazy<JsonSerializerSettings>(
+                () =>
+                {
+                    var settings = new JsonSerializerSettings
+                    {
+                        DateParseHandling = DateParseHandling.DateTimeOffset,
+                        MaxDepth = DefaultMaxDepth,
+                        ContractResolver = DefaultContractResolver,
+                        NullValueHandling = NullValueHandling.Ignore,
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: I don't like ignoring null values, it makes things harder to debug.

File: src/ProspaAspNetCoreApiNsb/Infrastructure/Behaviours/LoggingBehavior.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+        public async Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
+        {
+            var traceId = Guid.NewGuid();
+
+            try
+            {
+                _logger.Verbose("Handling {TRequest}. TraceId: {TraceId} {UtcNow}", typeof(TRequest).FullName, traceId, DateTime.UtcNow);
+
+                var response = await next();
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/Infrastructure/Behaviours/LoggingBehavior.cs]: if a handler throws an exception why don't we just let it bubble up and have it logged there?
This will lead to double logging exceptions.
We already have exception handling middleware.

File: src/ProspaAspNetCoreApiNsb/ConfigureOptions/FormattersMvcOptionsSetup.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.Extensions.Options;
+
+namespace ProspaAspNetCoreApiNsb.ConfigureOptions
+{
+    public class FormattersMvcOptionsSetup : IConfigureOptions<MvcOptions>
+    {
+        /// <inheritdoc />
+        public void Configure(MvcOptions options)
+        {
+            // Not useful for an API serving JSON.
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/ConfigureOptions/FormattersMvcOptionsSetup.cs]: can we just define this inline?

File: src/ProspaAspNetCoreApiNsb/ConfigureOptions/ConventionsMvcOptionsSetup.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.AspNetCore.Mvc;
+
+// ReSharper disable CheckNamespace
+namespace Microsoft.Extensions.Options
+    // ReSharper restore CheckNamespace
+{
+    public class ConventionsMvcOptionsSetup : IConfigureOptions<MvcOptions>
+    {
+        /// <inheritdoc />
+        public void Configure(MvcOptions options)
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApiNsb/ConfigureOptions/ConventionsMvcOptionsSetup.cs]: what's the purpose of this class?

File: src/ProspaAspNetCoreApi/Program.cs [hunk 10]
```diff
@@ -34,11 +34,7 @@ public static int Main(string[] args)
             Host.CreateDefaultBuilder(args)
                 .ConfigureAppConfiguration((context, builder) =>
                 {
-                    if (!string.IsNullOrEmpty(Constants.KeyVaultName))
-                    {
-                        var keyVaultEndpoint = $"https://{Constants.Environments.Prefix()}{Constants.KeyVaultName}.vault.azure.net/";
-                        builder.AddAzureKeyVault(keyVaultEndpoint);
-                    }
+                    builder.AddDefaultSources(args);
```
Comment (chinwobble) [inline src/ProspaAspNetCoreApi/Program.cs]: why was this changed?

File: docs/AspNetCoreApiAzureNsb.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+#### Configure Azure Keyvault
+
+The template is configured to add Azure Keyvault as a configuration provider. You'll need to create the KeyVault resource for the application and use MSI to connect.
+
+```
+keyvaultName="mykeyvaultName" # i.e. The name provided on template creation
+az keyvault create -n $keyvaultName -g $rg -l $location
+az keyvault secret set --vault-name $keyvaultName -n "EndpointKey" --value "secret"
+az keyvault secret set --vault-name $keyvaultName -n "ConnectionStrings--ServiceBus" --value $serviceBusConnectionString
```
Comment (chinwobble) [inline docs/AspNetCoreApiAzureNsb.md]: I think we should use arm templates to set the secret in keyvault instead of doing it manually with the CLI.
This is relevant since we have `p3disposable` which may delete certain resources.

File: README.md [hunk 12]
```diff
@@ -19,30 +19,10 @@ dotnet new list
...
-dotnet sln add .\src\MyNew.API\MyNew.API.csproj
-```
-
-> The template is configured to add secrets from Azure Keyvault using MSI. Before running the application you'll first need to either:
-> 1. Install the [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest) and then run `az login` OR
-> 2. Disable this by removing the `keyvaultName` key from `appsettings.{env}.json`
+|Short Name|Description|
+|------|--------|
+|[prospaapiazure](docs/AspNetCoreApiAzure.md)|An ASP.NET Core Api targetted at hosting on Azure|
+|[prospaapiazurensb](docs/AspNetCoreApiAzureNsb.md)|An ASP.NET Core Api with NServiceBus integration targetted at hosting on Azure|
```
Comment (chinwobble) [inline README.md]: You should update your VIM configuration to automatically format this table 😄

File: src/ProspaAspNetCoreApiNsb/Startup.Versioning.cs (hunk 2 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Startup.Versioning.cs]: removed, just using `AddApiVersioning`

File: src/ProspaAspNetCoreApiNsb/docs.json (hunk 1 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/docs.json]: No it doesn't, I'd like to eventually build a dynamic docs dashboard using this though

File: src/ProspaAspNetCoreApiNsb/Startup.SecurityHeaders.cs (hunk 3 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Startup.SecurityHeaders.cs]: removed

File: src/ProspaAspNetCoreApiNsb/Startup.Mvc.cs (hunk 4 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Startup.Mvc.cs]: I prefer to leave the `Startup.cs` succinct. I prefer to have explicit files for such configuration.

In this case I've just removed those configuration options completely in favor of just using MVC defaults

File: src/ProspaAspNetCoreApiNsb/Program.Configuration.cs (hunk 5 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Program.Configuration.cs]: Plan to start using Azure App Configuration to enable changes to config at runtime which would require similar code

File: src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs (hunk 6 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: This is only used for message serialization to service bus, I think it's fine and don't see why it makes things harder to debug. If values are null it's unnecessary to include them

File: src/ProspaAspNetCoreApiNsb/Infrastructure/Behaviours/LoggingBehavior.cs (hunk 7 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/Behaviours/LoggingBehavior.cs]: I've removed meditor logging behavior altogether

File: src/ProspaAspNetCoreApiNsb/ConfigureOptions/FormattersMvcOptionsSetup.cs (hunk 8 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/ConfigureOptions/FormattersMvcOptionsSetup.cs]: Again I prefer to have more succinct startup files. In this case I've removed the formatting options just keeping defaults

File: src/ProspaAspNetCoreApiNsb/ConfigureOptions/ConventionsMvcOptionsSetup.cs (hunk 9 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/ConfigureOptions/ConventionsMvcOptionsSetup.cs]: Left over from when modify mvcoptions in a previous version of the template

File: src/ProspaAspNetCoreApi/Program.cs (hunk 10 above)
Comment (alhardy) [reply src/ProspaAspNetCoreApi/Program.cs]: See last comment on the file

File: docs/AspNetCoreApiAzureNsb.md (hunk 11 above)
Comment (alhardy) [reply docs/AspNetCoreApiAzureNsb.md]: This is just a quick start, the template is arm agnostic

File: src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs (hunk 6 above)
Comment (chinwobble) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: Consider the following example:

```json
{
   "eventType": null,
   "field1": "x",
   "field2": "x",
   "field3": "x",
}
```

If a client forgets to a set value that is required like eventType its much easier to see, I can simply use the find tool to check for its existence.
If null values are not serialiesd, I cannot use the find tool to easily figure out what someone has forgotten.
This is particular relevant for C# since record types are not supported and its common to have properties with optional getter / setters and developers can forget to set properties

Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: There should be application validation on such things

Comment (chinwobble) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: What's the benefit of excluding them?

Comment (alhardy) [reply src/ProspaAspNetCoreApiNsb/Infrastructure/DefaultMessageJsonSerializerSettings.cs]: unnecessary, smaller message size. 

"If a client forgets to a set value that is required like eventType its much easier to see, I can simply use the find tool to check for its existence." If we decided on binary serialization for reasons this would make it more difficult. Also if values are missing it's equivalent to them being explicitly shown as null, not no real difference.

Anyways not precious on this, I've updated to just leave the default.
